File: SWEA/1954 달팽이 숫자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Solution {
	
	static int n;
	static int[][] grid;
	
	public static boolean inRange(int x, int y) {
		return 0 <= x && x < n && 0 <= y && y < n;
	}
	
	public static void moving() {
		int[] dx = {-1, 0, 1, 0};
		int[] dy = {0, 1, 0, -1};
		int dirNum = 1;
		int x = 0;
		int y = -1;
		int nx, ny;
		for (int i = 1; i < n*n+1; i++) {
			nx = x + dx[dirNum];
			ny = y + dy[dirNum];
			if (!inRange(nx, ny) || grid[nx][ny] != 0) {
				dirNum = (dirNum + 1) % 4;
			}
			x = x + dx[dirNum];
			y = y + dy[dirNum];
			grid[x][y] = i;
		}
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		int T = Integer.parseInt(br.readLine());  // 테스트케이스 수
		for (int i = 1; i <= T; i++) {
			n = Integer.parseInt(br.readLine());  // 달팽이의 크기
			grid = new int[n][n];
			moving();
			System.out.println("#" + i);
			
			for (int j = 0; j < n; j++) {
				for (int j2 = 0; j2 < n; j2++) {
					System.out.printf("%d ", grid[j][j2]);
				}
				System.out.println();
			}
		}
		
		
	}

}

File: SWEA/1954 달팽이 숫자.py
def in_range(x, y):
    return 0 <= x < n and 0 <= y < n

def moving():
    dx, dy = [-1, 0, 1, 0], [0, 1, 0, -1]
    dir_num = 1
    x, y = 0, -1
    for i in range(1, n*n+1):
        nx, ny = x + dx[dir_num], y + dy[dir_num]
        if not in_range(nx, ny) or grid[nx][ny] != 0:
            dir_num = (dir_num + 1) % 4
        x, y = x + dx[dir_num], y + dy[dir_num]
        grid[x][y] = i


T = int(input())
for t in range(1, T+1):
    n = int(input())
    grid = [[0] * n for _ in range(n)]
    moving()
    print(f"#{t}")
    for i in range(n):
        print(*grid[i])
